File: src/android/CSSInjector.java
package com.vnkhoado.cordova.changeappinfo;

import android.graphics.Color;
import android.os.Handler;
import android.os.Looper;
import android.util.Base64;
import android.webkit.WebResourceRequest;
import android.webkit.WebResourceResponse;
import android.webkit.WebView;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.engine.SystemWebView;
import org.apache.cordova.engine.SystemWebViewClient;
import org.apache.cordova.engine.SystemWebViewEngine;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;

public class CSSInjector extends CordovaPlugin {

    private static final String TAG = "CSSInjector";
    private static final String CSS_FILE_PATH = "www/assets/cdn-styles.css";
    private static final String CONFIG_FILE_PATH = "www/cordova-build-config.json";
    
    private String cachedCSS = null;
    private JSONObject cachedConfig = null;
    private Handler handler;
    private String backgroundColor = null;
    private boolean initialInjectionDone = false;
    private boolean isInjecting = false;
    private boolean isFirstPageLoad = true;
    private String configScript = null;

    @Override
    public void pluginInitialize() {
        super.pluginInitialize();
        
        // Read WEBVIEW_BACKGROUND_COLOR from preferences
        String bgColor = preferences.getString("WEBVIEW_BACKGROUND_COLOR", null);
        if (bgColor == null || bgColor.isEmpty()) {
            bgColor = preferences.getString("BackgroundColor", null);
        }
        if (bgColor == null || bgColor.isEmpty()) {
            bgColor = preferences.getString("SplashScreenBackgroundColor", null);
        }
        
        // Default to white if no color specified
        if (bgColor == null || bgColor.isEmpty()) {
            bgColor = "#FFFFFF";
        }
        
        backgroundColor = bgColor;
        
        // Set WebView and Activity background
        final String finalBgColor = bgColor;
        cordova.getActivity().runOnUiThread(() -> {
            try {
                int color = parseHexColor(finalBgColor);
                cordova.getActivity().getWindow().setBackgroundDrawable(
                    new android.graphics.drawable.ColorDrawable(color)
                );
                cordova.getActivity().getWindow().getDecorView().setBackgroundColor(color);
                if (webView != null && webView.getView() != null) {
                    webView.getView().setBackgroundColor(color);
                }
                android.util.Log.d(TAG, "Background set to: " + finalBgColor);
            } catch (IllegalArgumentException e) {
                android.util.Log.e(TAG, "Invalid color: " + finalBgColor, e);
            }
        });
        
        // Pre-load CSS and config
        cachedCSS = readCSSFromAssets();
        cachedConfig = readConfigFromAssets();
        
        // Pre-build config script for injection
        buildConfigScript();
        
        handler = new Handler(Looper.getMainLooper());
        
        // Setup WebViewClient to listen for page loads
        setupWebViewClient();
        
        // FIX: Pre-inject CSS early to ensure it's ready
        handler.postDelayed(() -> {
            android.util.Log.d(TAG, "Early CSS pre-injection");
            injectBackgroundColorCSS(backgroundColor);
            injectCSSIntoWebView();
        }, 100);
        
        android.util.Log.d(TAG, "CSSInjector initialized with background: " + backgroundColor);
    }

    /**
     * Build config script that will be injected into HTML
     */
    private void buildConfigScript() {
        try {
            JSONObject config = cachedConfig;
            if (config == null) {
                config = readConfigFromAssets();
                cachedConfig = config;
            }
            
            if (config == null) {
                android.util.Log.w(TAG, "No config found for script building");
                return;
            }
            
            // Add background color to config
            if (backgroundColor != null && !backgroundColor.isEmpty()) {
                config.put("backgroundColor", backgroundColor);
            }
            
            String configJSON = config.toString();
            
            // Build inline script
            configScript = "<script type='text/javascript'>" +
                "(function(){" +
                "try{" +
                "var config=" + configJSON + ";" +
                "window.CORDOVA_BUILD_CONFIG=config;" +
                "window.AppConfig=config;" +
                "console.log('[Native-Inline] Config injected:',config);" +
                "}catch(e){" +
                "console.error('[Native-Inline] Config failed:',e);" +
                "}" +
                "})();" +
                "</script>";
            
            android.util.Log.d(TAG, "Config script built successfully");
        } catch (Exception e) {
            android.util.Log.e(TAG, "Failed to build config script", e);
        }
    }

    /**
     * Setup WebViewClient to intercept page load events
     */

    private void setupWebViewClient() {
        cordova.getActivity().runOnUiThread(() -> {
            try {
                if (webView != null && webView.getView() instanceof SystemWebView) {
                    SystemWebView systemWebView = (SystemWebView) webView.getView();

                    // Get the engine
                    SystemWebViewEngine engine = (SystemWebViewEngine) webView.getEngine();

                    // Create custom SystemWebViewClient
                    SystemWebViewClient customClient = new SystemWebViewClient(engine) {
                        
                        @Override
                        public WebResourceResponse shouldInterceptRequest(WebView view, WebResourceRequest request) {
                            String url = request.getUrl().toString();
                            
                            // FIX: Intercept HTML pages to inject config BEFORE any JS runs
                            if (url.endsWith("index.html") || url.contains("StaffPortalMobile")) {
                                try {
                                    android.util.Log.d(TAG, "Intercepting: " + url);
                                    
                                    // Get original HTML from cache/network
                                    WebResourceResponse response = super.shouldInterceptRequest(view, request);
                                    if (response != null && response.getData() != null) {
                                        // Read original HTML
                                        String html = readStream(response.getData());
                                        
                                        // Inject config script at the very beginning of <head>
                                        if (configScript != null && html.contains("<head>")) {
                                            html = html.replace("<head>", "<head>" + configScript);
                                            android.util.Log.d(TAG, "Config injected into HTML head");
                                        }
                                        
                                        // Return modified HTML
                                        return new WebResourceResponse(
                                            "text/html",
                                            "UTF-8",
                                            new ByteArrayInputStream(html.getBytes(StandardCharsets.UTF_8))
                                        );
                                    }
                                } catch (Exception e) {
                                    android.util.Log.e(TAG, "Failed to intercept HTML", e);
                                }
                            }
                            
                            return super.shouldInterceptRequest(view, request);
                        }
                        
                        @Override
                        public void onPageStarted(WebView view, String url, android.graphics.Bitmap favicon) {
                            super.onPageStarted(view, url, favicon);
                            android.util.Log.d(TAG, "Page started: " + url);

                            // FIX: Set native background immediately
                            if (backgroundColor != null && !backgroundColor.isEmpty()) {
                                try {
                                    int color = parseHexColor(backgroundColor);
                                    view.setBackgroundColor(color);
                                    
                                    // FIX: Inject background CSS immediately when page starts
                                    injectBackgroundColorCSS(backgroundColor);
                                } catch (Exception e) {
                                    android.util.Log.e(TAG, "Failed to set bg on page start", e);
                                }
                            }
                            
                            // FIX: Inject config IMMEDIATELY via JavaScript as backup
                            injectBuildConfig();
                            
                            // FIX: For first page load, inject aggressively
                            if (isFirstPageLoad) {
                                android.util.Log.d(TAG, "First page load - aggressive injection");
                                injectAllContent();
                                isFirstPageLoad = false;
                            }
                        }

                        @Override
                        public void onPageFinished(WebView view, String url) {
                            super.onPageFinished(view, url);
                            android.util.Log.d(TAG, "Page finished: " + url);

                            // FIX: Inject immediately without delay
                            injectAllContent();
                        }
                    };

                    systemWebView.setWebViewClient(customClient);
                    android.util.Log.d(TAG, "Custom SystemWebViewClient installed");
                }
            } catch (Exception e) {
                android.util.Log.e(TAG, "Failed to setup WebViewClient", e);
            }
        });
    }

    /**
     * Read InputStream to String
     */
    private String readStream(InputStream is) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(is, StandardCharsets.UTF_8));
        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            sb.append(line).append("\n");
        }
        return sb.toString();
    }

    @Override
    public void onResume(boolean multitasking) {
        super.onResume(multitasking);
        
        // FIX: Inject immediately on first resume
        if (!initialInjectionDone) {
            injectAllContent();
            initialInjectionDone = true;
            android.util.Log.d(TAG, "onResume - immediate injection");
        }
        
        android.util.Log.d(TAG, "onResume");
    }

    /**
     * Inject all content: config, background CSS, CDN CSS
     */
    private void injectAllContent() {
        // Prevent duplicate injections
        if (isInjecting) {
            android.util.Log.d(TAG, "Already injecting, skipping...");
            return;
        }
        
        isInjecting = true;
        
        try {
            // 1. Inject build config into window FIRST (most important)
            injectBuildConfig();
            
            // 2. Inject background color CSS immediately
            if (backgroundColor != null && !backgroundColor.isEmpty()) {
                injectBackgroundColorCSS(backgroundColor);
            }
            
            // 3. Inject CDN CSS with minimal delay
            handler.postDelayed(() -> injectCSSIntoWebView(), 50);
            
            android.util.Log.d(TAG, "All content injection scheduled");
        } finally {
            // Reset flag after a delay
            handler.postDelayed(() -> {
                isInjecting = false;
            }, 300);
        }
    }

    /**
     * Inject build config from JSON file into window variable
     */
    private void injectBuildConfig() {
        cordova.getActivity().runOnUiThread(() -> {
            try {
                JSONObject config = cachedConfig;
                if (config == null) {
                    config = readConfigFromAssets();
                    cachedConfig = config;
                }
                
                if (config == null) {
                    android.util.Log.w(TAG, "No config found, skipping injection");
                    return;
                }
                
                // Add background color to config
                if (backgroundColor != null && !backgroundColor.isEmpty()) {
                    config.put("backgroundColor", backgroundColor);
                }
                
                // Convert to JSON string and escape for JavaScript
                String configJSON = config.toString();
                String escapedJSON = configJSON
                    .replace("\\", "\\\\")
                    .replace("'", "\\'")
                    .replace("\"", "\\\"")
                    .replace("\n", "\\n");
                
                CordovaWebView cordovaWebView = this.webView;
                if (cordovaWebView != null) {
                    String javascript = "(function() {" +
                        "  try {" +
                        "    var config = JSON.parse(\"" + escapedJSON + "\");" +
                        "    window.CORDOVA_BUILD_CONFIG = config;" +
                        "    window.AppConfig = config;" +
                        "    console.log('[Native-JS] Build config injected:', config);" +
                        "    " +
                        "    if (typeof CustomEvent !== 'undefined') {" +
                        "      window.dispatchEvent(new CustomEvent('cordova-config-ready', { detail: config }));" +
                        "    }" +
                        "  } catch(e) {" +
                        "    console.error('[Native-JS] Config injection failed:', e);" +
                        "  }" +
                        "})();";
                    
                    cordovaWebView.loadUrl("javascript:" + javascript);
                    android.util.Log.d(TAG, "Build config injected via JavaScript");
                }
            } catch (Exception e) {
                android.util.Log.e(TAG, "Failed to inject build config", e);
            }
        });
    }

    /**
     * Read config JSON from assets
     */
    private JSONObject readConfigFromAssets() {
        StringBuilder content = new StringBuilder();
        
        try {
            InputStream inputStream = cordova.getActivity().getAssets().open(CONFIG_FILE_PATH);
            BufferedReader reader = new BufferedReader(
                new InputStreamReader(inputStream, StandardCharsets.UTF_8)
            );
            
            String line;
            while ((line = reader.readLine()) != null) {
                content.append(line);
            }
            
            reader.close();
            inputStream.close();
            
            return new JSONObject(content.toString());
            
        } catch (IOException e) {
            android.util.Log.w(TAG, "Config file not found: " + CONFIG_FILE_PATH);
            return null;
        } catch (JSONException e) {
            android.util.Log.e(TAG, "Failed to parse config JSON", e);
            return null;
        }
    }

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        if (action.equals("injectCSS")) {
            injectCSSIntoWebView();
            callbackContext.success("CSS injected");
            return true;
        } else if (action.equals("getConfig")) {
            // Allow JS to get config on demand
            JSONObject config = cachedConfig != null ? cachedConfig : readConfigFromAssets();
            if (config != null) {
                callbackContext.success(config);
            } else {
                callbackContext.error("Config not available");
            }
            return true;
        } else if (action.equals("injectBackground")) {
            // Allow JS to manually trigger background injection
            if (backgroundColor != null && !backgroundColor.isEmpty()) {
                injectBackgroundColorCSS(backgroundColor);
                callbackContext.success("Background injected: " + backgroundColor);
            } else {
                callbackContext.error("No background color configured");
            }
            return true;
        }
        return false;
    }

    private int parseHexColor(String hexColor) throws IllegalArgumentException {
        String hex = hexColor.trim();
        if (hex.startsWith("#")) {
            hex = hex.substring(1);
        }
        if (hex.length() != 6 && hex.length() != 8) {
            throw new IllegalArgumentException("Hex must be 6 or 8 chars");
        }
        try {
            if (hex.length() == 6) {
                return Color.parseColor("#FF" + hex);
            } else {
                return Color.parseColor("#" + hex);
            }
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException("Invalid color: " + hexColor, e);
        }
    }

    /**
     * Inject background color CSS - OPTIMIZED VERSION
     * Single injection, no repetition
     */
    private void injectBackgroundColorCSS(final String bgColor) {
        cordova.getActivity().runOnUiThread(() -> {
            try {
                // Set native WebView background FIRST
                if (webView != null && webView.getView() != null) {
                    try {
                        int color = parseHexColor(bgColor);
                        webView.getView().setBackgroundColor(color);
                    } catch (Exception e) {
                        android.util.Log.e(TAG, "Failed to set native bg", e);
                    }
                }
                
                // Then inject CSS
                CordovaWebView cordovaWebView = this.webView;
                if (cordovaWebView != null) {
                    // CSS with multiple selectors for better coverage
                    String css = "html, body, #root, #app, .app-container, .screen, .page-wrapper { " +
                        "background-color: " + bgColor + " !important; " +
                        "background: " + bgColor + " !important; " +
                        "margin: 0; padding: 0; " +
                        "}";
                    
                    String javascript = "(function() {" +
                        "  try {" +
                        "    if (document.documentElement) {" +
                        "      document.documentElement.style.backgroundColor = '" + bgColor + "';" +
                        "    }" +
                        "    if (document.body) {" +
                        "      document.body.style.backgroundColor = '" + bgColor + "';" +
                        "    }" +
                        "    " +
                        "    var target = document.head || document.getElementsByTagName('head')[0];" +
                        "    if (target) {" +
                        "      var s = document.getElementById('cordova-bg');" +
                        "      if (!s) {" +
                        "        s = document.createElement('style');" +
                        "        s.id = 'cordova-bg';" +
                        "        s.textContent = '" + css.replace("'", "\\'") + "';" +
                        "        target.insertBefore(s, target.firstChild);" +
                        "        console.log('[Native] Background CSS: " + bgColor + "');" +
                        "      }" +
                        "    }" +
                        "  } catch(e) { console.error('[Native] BG failed:', e); }" +
                        "})();";
                    
                    cordovaWebView.loadUrl("javascript:" + javascript);
                }
            } catch (Exception e) {
                android.util.Log.e(TAG, "Background CSS failed", e);
            }
        });
    }

    private void injectCSSIntoWebView() {
        cordova.getActivity().runOnUiThread(() -> {
            try {
                String cssContent = cachedCSS;
                if (cssContent == null || cssContent.isEmpty()) {
                    cssContent = readCSSFromAssets();
                    cachedCSS = cssContent;
                }
                
                if (cssContent != null && !cssContent.isEmpty()) {
                    CordovaWebView cordovaWebView = this.webView;
                    if (cordovaWebView != null) {
                        String javascript = buildCSSInjectionScript(cssContent);
                        cordovaWebView.loadUrl("javascript:" + javascript);
                        android.util.Log.d(TAG, "CDN CSS injected (" + cssContent.length() + " bytes)");
                    }
                }
            } catch (Exception e) {
                android.util.Log.e(TAG, "CSS injection failed", e);
            }
        });
    }

    private String readCSSFromAssets() {
        StringBuilder cssContent = new StringBuilder();
        try {
            InputStream inputStream = cordova.getActivity().getAssets().open(CSS_FILE_PATH);
            BufferedReader reader = new BufferedReader(
                new InputStreamReader(inputStream, StandardCharsets.UTF_8)
            );
            String line;
            while ((line = reader.readLine()) != null) {
                cssContent.append(line).append("\n");
            }
            reader.close();
            inputStream.close();
        } catch (IOException e) {
            android.util.Log.e(TAG, "Failed to read CSS", e);
            return null;
        }
        return cssContent.toString();
    }

    private String buildCSSInjectionScript(String cssContent) {
        try {
            byte[] cssBytes = cssContent.getBytes(StandardCharsets.UTF_8);
            String base64CSS = Base64.encodeToString(cssBytes, Base64.NO_WRAP);
            
            return "(function() {" +
                   "  function inject() {" +
                   "    try {" +
                   "      var target = document.head || document.getElementsByTagName('head')[0] || document.documentElement;" +
                   "      if (!target) {" +
                   "        setTimeout(inject, 100);" +
                   "        return;" +
                   "      }" +
                   "      if (!document.getElementById('cdn-styles')) {" +
                   "        var b64 = '" + base64CSS + "';" +
                   "        var css = decodeURIComponent(escape(atob(b64)));" +
                   "        var s = document.createElement('style');" +
                   "        s.id = 'cdn-styles';" +
                   "        s.textContent = css;" +
                   "        target.appendChild(s);" +
                   "        console.log('[Native] CDN CSS loaded');" +
                   "      }" +
                   "    } catch(e) { console.error('[Native] CDN CSS failed:', e); }" +
                   "  }" +
                   "  if (document.readyState === 'loading') {" +
                   "    document.addEventListener('DOMContentLoaded', inject);" +
                   "  } else {" +
                   "    inject();" +
                   "  }" +
                   "})();";
        } catch (Exception e) {
            return buildFallbackInjectionScript(cssContent);
        }
    }

    private String buildFallbackInjectionScript(String cssContent) {
        String escaped = cssContent
            .replace("\\", "\\\\")
            .replace("'", "\\'")
            .replace("\"", "\\\"")
            .replace("\n", "\\n")
            .replace("\r", "")
            .replace("\t", "\\t");
        
        return "(function() {" +
               "  function inject() {" +
               "    try {" +
               "      var target = document.head || document.getElementsByTagName('head')[0] || document.documentElement;" +
               "      if (!target) {" +
               "        setTimeout(inject, 100);" +
               "        return;" +
               "      }" +
               "      if (!document.getElementById('cdn-styles')) {" +
               "        var s = document.createElement('style');" +
               "        s.id = 'cdn-styles';" +
               "        s.textContent = '" + escaped + "';" +
               "        target.appendChild(s);" +
               "        console.log('[Native] CDN CSS loaded');" +
               "      }" +
               "    } catch(e) { console.error('[Native] CSS failed:', e); }" +
               "  }" +
               "  if (document.readyState === 'loading') {" +
               "    document.addEventListener('DOMContentLoaded', inject);" +
               "  } else {" +
               "    inject();" +
               "  }" +
               "})();";
    }
}

File: src/ios/CSSInjector.swift
import Foundation
import WebKit
import UIKit

@objc(CSSInjector)
class CSSInjector: CDVPlugin {
    
    private static let CSS_FILE_PATH = "www/assets/cdn-styles.css"
    private static let CONFIG_FILE_PATH = "www/cordova-build-config.json"
    private var cachedCSS: String?
    private var cachedConfig: [String: Any]?
    
    override func pluginInitialize() {
        super.pluginInitialize()
        
        // Read WEBVIEW_BACKGROUND_COLOR from preferences (with fallbacks)
        let bgColor = getBackgroundColor()
        
        if let color = bgColor {
            setWebViewBackgroundColor(colorString: color)
        }
        
        // Pre-load CSS and config
        cachedCSS = readCSSFromBundle()
        cachedConfig = readConfigFromBundle()
        
        // ⭐ KEY FIX: Install WKUserScripts immediately
        // This ensures CSS/config inject BEFORE page loads (no race condition)
        installUserScripts()
        
        print("[CSSInjector] Plugin initialized with WKUserScript injection")
    }
    
    // MARK: - WKUserScript Installation (Option C - Best Practice)
    
    /**
     * Install WKUserScripts for CSS and Config injection
     * Scripts run at .atDocumentStart = BEFORE page renders
     * Eliminates timing issues on fresh app install
     */
    private func installUserScripts() {
        DispatchQueue.main.async {
            guard let wkWebView = self.webView as? WKWebView else {
                print("[CSSInjector] WebView not available")
                return
            }
            
            let contentController = wkWebView.configuration.userContentController
            
            // 1. Install Config UserScript (highest priority)
            if let configScript = self.buildConfigUserScript() {
                contentController.addUserScript(configScript)
                print("[CSSInjector] ✅ Config UserScript installed")
            }
            
            // 2. Install Background Color UserScript
            if let bgColor = self.getBackgroundColor() {
                let bgScript = self.buildBackgroundUserScript(color: bgColor)
                contentController.addUserScript(bgScript)
                print("[CSSInjector] ✅ Background UserScript installed: \(bgColor)")
            }
            
            // 3. Install CSS UserScript
            if let cssScript = self.buildCSSUserScript() {
                contentController.addUserScript(cssScript)
                if let cssSize = self.cachedCSS?.count {
                    print("[CSSInjector] ✅ CSS UserScript installed (\(cssSize) bytes)")
                }
            }
            
            print("[CSSInjector] All UserScripts installed successfully")
        }
    }
    
    /**
     * Build Config injection UserScript
     * Injects window.CORDOVA_BUILD_CONFIG before page loads
     */
    private func buildConfigUserScript() -> WKUserScript? {
        guard var configDict = cachedConfig else {
            print("[CSSInjector] No config to inject")
            return nil
        }
        
        // Add background color to config
        if let bgColor = getBackgroundColor() {
            configDict["backgroundColor"] = bgColor
        }
        
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: configDict, options: [])
            guard let jsonString = String(data: jsonData, encoding: .utf8) else { return nil }
            
            let escapedJSON = jsonString
                .replacingOccurrences(of: "\\", with: "\\\\")
                .replacingOccurrences(of: "'", with: "\\'")
                .replacingOccurrences(of: "\"", with: "\\\"")
                .replacingOccurrences(of: "\n", with: "\\n")
            
            let javascript = """
            (function() {
                try {
                    var config = JSON.parse("\(escapedJSON)");
                    window.CORDOVA_BUILD_CONFIG = config;
                    window.AppConfig = config;
                    console.log('[Native iOS UserScript] Config injected at document start');
                    
                    // Dispatch event when DOM is ready
                    if (document.readyState === 'loading') {
                        document.addEventListener('DOMContentLoaded', function() {
                            if (typeof CustomEvent !== 'undefined') {
                                window.dispatchEvent(new CustomEvent('cordova-config-ready', { detail: config }));
                            }
                        });
                    } else {
                        if (typeof CustomEvent !== 'undefined') {
                            window.dispatchEvent(new CustomEvent('cordova-config-ready', { detail: config }));
                        }
                    }
                } catch(e) {
                    console.error('[Native iOS UserScript] Config injection failed:', e);
                }
            })();
            """
            
            // ⭐ atDocumentStart = inject BEFORE page loads (key to fix timing issue)
            return WKUserScript(
                source: javascript,
                injectionTime: .atDocumentStart,
                forMainFrameOnly: true
            )
        } catch {
            print("[CSSInjector] Failed to build config script: \(error)")
            return nil
        }
    }
    
    /**
     * Build Background Color UserScript
     * Sets background color before page renders (prevents white flash)
     */
    private func buildBackgroundUserScript(color: String) -> WKUserScript {
        let css = "html, body, #root, #app, .app-container { background-color: \(color) !important; background: \(color) !important; margin: 0; padding: 0; }"
        let escapedCSS = css.replacingOccurrences(of: "'", with: "\\'")
        
        let javascript = """
        (function() {
            try {
                // Set inline styles immediately
                if (document.documentElement) {
                    document.documentElement.style.backgroundColor = '\(color)';
                }
                
                // Create style tag
                var style = document.createElement('style');
                style.id = 'cordova-bg-color';
                style.textContent = '\(escapedCSS)';
                (document.head || document.documentElement).appendChild(style);
                
                console.log('[Native iOS UserScript] Background color injected: \(color)');
            } catch(e) {
                console.error('[Native iOS UserScript] Background injection failed:', e);
            }
        })();
        """
        
        return WKUserScript(
            source: javascript,
            injectionTime: .atDocumentStart,
            forMainFrameOnly: true
        )
    }
    
    /**
     * Build CSS injection UserScript
     * Injects CDN CSS before page renders
     */
    private func buildCSSUserScript() -> WKUserScript? {
        guard let css = cachedCSS, !css.isEmpty else {
            print("[CSSInjector] No CSS to inject")
            return nil
        }
        
        guard let base64CSS = encodeToBase64(cssContent: css) else {
            print("[CSSInjector] Failed to encode CSS to Base64")
            return buildFallbackCSSUserScript(cssContent: css)
        }
        
        let javascript = """
        (function() {
            try {
                if (!document.getElementById('cdn-injected-styles')) {
                    var base64CSS = '\(base64CSS)';
                    var decodedCSS = decodeURIComponent(escape(atob(base64CSS)));
                    var style = document.createElement('style');
                    style.id = 'cdn-injected-styles';
                    style.textContent = decodedCSS;
                    (document.head || document.documentElement).appendChild(style);
                    console.log('[Native iOS UserScript] CDN CSS injected (\(css.count) bytes)');
                }
            } catch(e) {
                console.error('[Native iOS UserScript] CSS injection failed:', e);
            }
        })();
        """
        
        // ⭐ atDocumentStart = CSS ready BEFORE page renders
        return WKUserScript(
            source: javascript,
            injectionTime: .atDocumentStart,
            forMainFrameOnly: true
        )
    }
    
    /**
     * Fallback CSS UserScript (if Base64 encoding fails)
     */
    private func buildFallbackCSSUserScript(cssContent: String) -> WKUserScript? {
        let escapedCSS = cssContent
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "'", with: "\\'")
            .replacingOccurrences(of: "\"", with: "\\\"")
            .replacingOccurrences(of: "\n", with: "\\n")
            .replacingOccurrences(of: "\r", with: "")
            .replacingOccurrences(of: "\t", with: "\\t")
        
        let javascript = """
        (function() {
            try {
                if (!document.getElementById('cdn-injected-styles')) {
                    var style = document.createElement('style');
                    style.id = 'cdn-injected-styles';
                    style.textContent = '\(escapedCSS)';
                    (document.head || document.documentElement).appendChild(style);
                    console.log('[Native iOS UserScript] CSS injected (fallback method)');
                }
            } catch(e) {
                console.error('[Native iOS UserScript] CSS injection failed:', e);
            }
        })();
        """
        
        return WKUserScript(
            source: javascript,
            injectionTime: .atDocumentStart,
            forMainFrameOnly: true
        )
    }
    
    // MARK: - Plugin Methods (for manual JS calls)
    
    @objc(injectCSS:)
    func injectCSS(command: CDVInvokedUrlCommand) {
        // Manual injection via JS call (fallback)
        injectCSSViaEvaluateJavaScript()
        
        let pluginResult = CDVPluginResult(
            status: CDVCommandStatus_OK,
            messageAs: "CSS injected"
        )
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    @objc(getConfig:)
    func getConfig(command: CDVInvokedUrlCommand) {
        var config = cachedConfig
        if config == nil {
            config = readConfigFromBundle()
            cachedConfig = config
        }
        
        if let configDict = config {
            let pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: configDict
            )
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        } else {
            let pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Config not available"
            )
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
    }
    
    // MARK: - Helper Methods
    
    /**
     * Get background color from preferences (with fallbacks)
     */
    private func getBackgroundColor() -> String? {
        if let color = self.commandDelegate.settings["webview_background_color"] as? String {
            return color
        } else if let color = self.commandDelegate.settings["backgroundcolor"] as? String {
            return color
        } else if let color = self.commandDelegate.settings["splashscreenbackgroundcolor"] as? String {
            return color
        }
        return nil
    }
    
    /**
     * Read config JSON from bundle
     */
    private func readConfigFromBundle() -> [String: Any]? {
        guard let bundlePath = Bundle.main.path(forResource: "www", ofType: nil) else {
            print("[CSSInjector] www bundle path not found")
            return nil
        }
        
        let configPath = (bundlePath as NSString).appendingPathComponent("cordova-build-config.json")
        
        do {
            let jsonData = try Data(contentsOf: URL(fileURLWithPath: configPath))
            let config = try JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any]
            return config
        } catch {
            print("[CSSInjector] Failed to read config: \(error.localizedDescription)")
            return nil
        }
    }
    
    /**
     * Read CSS content from bundle www/assets/cdn-styles.css with UTF-8 encoding
     */
    private func readCSSFromBundle() -> String? {
        guard let bundlePath = Bundle.main.path(forResource: "www", ofType: nil) else {
            print("[CSSInjector] www bundle path not found")
            return nil
        }
        
        let cssPath = (bundlePath as NSString).appendingPathComponent("assets/cdn-styles.css")
        
        do {
            let cssContent = try String(contentsOfFile: cssPath, encoding: .utf8)
            return cssContent
        } catch {
            print("[CSSInjector] Failed to read CSS file: \(error.localizedDescription)")
            return nil
        }
    }
    
    /**
     * Encode CSS content to Base64
     */
    private func encodeToBase64(cssContent: String) -> String? {
        guard let data = cssContent.data(using: .utf8) else {
            print("[CSSInjector] Failed to encode CSS to UTF-8")
            return nil
        }
        
        return data.base64EncodedString(options: [])
    }
    
    // MARK: - WebView Background Color (Native)
    
    /**
     * Set WebView background color to prevent white flash
     */
    private func setWebViewBackgroundColor(colorString: String) {
        DispatchQueue.main.async {
            guard let webView = self.webView as? WKWebView else {
                print("[CSSInjector] WebView not available for background color")
                return
            }
            
            // Parse hex color
            if let color = self.hexStringToUIColor(hex: colorString) {
                webView.backgroundColor = color
                webView.isOpaque = false
                webView.scrollView.backgroundColor = color
                print("[CSSInjector] Native WebView background set to: \(colorString)")
            } else {
                // Fallback to clear
                webView.backgroundColor = .clear
                webView.isOpaque = false
                print("[CSSInjector] Invalid color format, using clear: \(colorString)")
            }
        }
    }
    
    /**
     * Convert hex string to UIColor
     */
    private func hexStringToUIColor(hex: String) -> UIColor? {
        var hexSanitized = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        hexSanitized = hexSanitized.replacingOccurrences(of: "#", with: "")
        
        var rgb: UInt64 = 0
        
        guard Scanner(string: hexSanitized).scanHexInt64(&rgb) else {
            return nil
        }
        
        let length = hexSanitized.count
        
        if length == 6 {
            let r = CGFloat((rgb & 0xFF0000) >> 16) / 255.0
            let g = CGFloat((rgb & 0x00FF00) >> 8) / 255.0
            let b = CGFloat(rgb & 0x0000FF) / 255.0
            return UIColor(red: r, green: g, blue: b, alpha: 1.0)
        } else if length == 8 {
            let r = CGFloat((rgb & 0xFF000000) >> 24) / 255.0
            let g = CGFloat((rgb & 0x00FF0000) >> 16) / 255.0
            let b = CGFloat((rgb & 0x0000FF00) >> 8) / 255.0
            let a = CGFloat(rgb & 0x000000FF) / 255.0
            return UIColor(red: r, green: g, blue: b, alpha: a)
        }
        
        return nil
    }
    
    // MARK: - Fallback: Manual Injection (via evaluateJavaScript)
    
    /**
     * Fallback CSS injection via evaluateJavaScript
     * Used only when called manually from JS or as backup
     */
    private func injectCSSViaEvaluateJavaScript() {
        DispatchQueue.main.async {
            guard let wkWebView = self.webView as? WKWebView else {
                print("[CSSInjector] WKWebView not available")
                return
            }
            
            var cssContent = self.cachedCSS
            if cssContent == nil || cssContent!.isEmpty {
                cssContent = self.readCSSFromBundle()
                self.cachedCSS = cssContent
            }
            
            guard let css = cssContent, !css.isEmpty else {
                print("[CSSInjector] CSS file not found or empty")
                return
            }
            
            guard let base64CSS = self.encodeToBase64(cssContent: css) else {
                print("[CSSInjector] Failed to encode CSS")
                return
            }
            
            let javascript = """
            (function() {
                try {
                    if (!document.getElementById('cdn-injected-styles')) {
                        var base64CSS = '\(base64CSS)';
                        var decodedCSS = decodeURIComponent(escape(atob(base64CSS)));
                        var style = document.createElement('style');
                        style.id = 'cdn-injected-styles';
                        style.textContent = decodedCSS;
                        (document.head || document.documentElement).appendChild(style);
                        console.log('[Native iOS Fallback] CSS injected');
                    }
                } catch(e) {
                    console.error('[Native iOS Fallback] CSS injection failed:', e);
                }
            })();
            """
            
            wkWebView.evaluateJavaScript(javascript) { (_, error) in
                if let error = error {
                    print("[CSSInjector] Fallback CSS injection failed: \(error.localizedDescription)")
                } else {
                    print("[CSSInjector] Fallback CSS injected successfully")
                }
            }
        }
    }
}
